proknow-sdk/Upload/Uploads.cs

```diff
@@ -85,83 +86,95 @@ public async Task<IList<UploadResult>> UploadAsync(string workspace, string path
+                        unresolvedUploadResults.RemoveAt(index);
+                    }
 
-                    // If one was found
-                    if (lastTerminalUploadProcessingResult != null)
+                    // If there are still unresolved uploads and possibly another page of results
+                    if (unresolvedUploadResults.Count > 0 && thisPageUploadProcessingResults.Count > 0)
                     {
-                        // Update the query parameters to search for processed uploads updated after this one
-                        if (queryParameters == null)
+                        // Get the last upload on the current page whose processing has reached a terminal status
+                        var lastTerminalUploadProcessingResult = thisPageUploadProcessingResults.LastOrDefault(t => _terminalStatuses.Contains(t.Status));
+
+                        // If one was found
+                        if (lastTerminalUploadProcessingResult != null)
                         {
-                            queryParameters = new Dictionary<string, object>();
+                            // Update the query parameters to search for processed uploads updated after this one
+                            if (queryParameters == null)
+                            {
+                                queryParameters = new Dictionary<string, object>();
+                            }
+                            queryParameters["updated"] = lastTerminalUploadProcessingResult.UpdatedAt;
+                            queryParameters["after"] = lastTerminalUploadProcessingResult.Id;
                         }
-                        queryParameters["updated"] = lastTerminalUploadProcessingResult.UpdatedAt;
-                        queryParameters["after"] = lastTerminalUploadProcessingResult.Id;
                     }
+                    else
+                    {
+                        break;
+                    }
+                } // End loop through pages of processing results
 
-                    // Give the uploads some time to process with each retry waiting twice as long as the previous retry
-                    await Task.Delay(nextRetryDelay);
-                    totalRetryDelay += nextRetryDelay;
-                    nextRetryDelay *= 2;
+                // If there are still unresolved uploads and retries have not been exhausted
+                if (unresolvedUploadResults.Count > 0 && retryDelayIndex < RetryDelays.Count)
+                {
+                    // Give the uploads some time to process
+                    await Task.Delay(RetryDelays[retryDelayIndex++]);
                 }
-            }
-
-            // Verify there are no unresolved uploads
-            if (unresolvedUploadResults.Count > 0)
-            {
-                throw new ProKnowException($"ProKnow processing of ${unresolvedUploadResults.Count} DICOM objects has not completed.  Timed out after {totalRetryDelay / 1000} sec.");

```

Comment:
I see that you're no longer throwing anymore. In the case where not all of the items are resolved in time due to, for example, a bug in ProKnow or a large backlog of items in the queue, would the user know that the timeout occurred?

Reply:
I'll change the return value to a structure that includes a flag indicating whether retry delays were exhausted.